File: projects/dodgem/scripts/Ball.js
// [1] soccer : bounces, affected by gravity downwards
// [2] beach ball : move diagonal, bounce top/bottom
// [3] spikey : move straight right, accelerates
// [4] bowling ball : like beach ball, slower
// [5] smiley : bounces, affected by gravity up/downwards

/* sounds
beach -> car door
bowling -> cannon
spike -> crunchy
smile -> drip
soccer -> switch1
*/

class Ball extends Entity {
    init () {
		this.addSprite("ball", {image:"balls", frames:5, speed:0, frame_size:[128,128]})
		this.sprite_pivot.set(this.sprite_width/2, this.sprite_height/2)
		
		let b_type = Util.rand_range(0,6);
		this.b_type = b_type;
		let scale;
		if (b_type == 4) // bowling ball
			scale = 2;
		else if (b_type == 4 || b_type == 2) // bowling ball, beach ball
			scale = 1;
		else 
			scale = 0.5;
		
		this.sprite_scale.set(scale);
		this.addShape('main',{type:'circle',shape:[0,0,this.sprite_width/2 - (15*scale)]})
		
		this.x = -this.sprite_width;
		this.y = Util.rand_range(0,Game.height/2);
		this.sprite_frame = b_type-1;
		this.debug = true;
		
		switch (b_type) {
			case 1: // soccer
				this.hspeed = 3;
				this.gravity_direction = 90;
				this.gravity = 0.2;
				break;
			case 2: // beach ball
				this.moveDirection(Util.rand_choose([-45,45]),10)
				break;
			case 3: // spikey
				this.hspeed = 0;
				break;
			case 4: // bowling ball
				this.moveDirection(Util.rand_choose([-45,45]),2)
				break;
			case 5: // smiley
				this.hspeed = 3;
				this.gravity_direction = 90 * -Util.rand_choose([-1,1]);
				this.gravity = 0.2;
				break;
		}
	}
    update (dt) {
		if (this.y < this.sprite_height/2) {
			this.y = this.sprite_height/2
			this.vspeed = -this.vspeed;
		}
		if (this.y > Game.height - this.sprite_height/2) {
			this.y = Game.height - this.sprite_height/2
			this.vspeed = -this.vspeed;
		}
		// beach ball
		if (this.b_type == 3)
			this.hspeed += 0.2;
		
		if (this.x > Game.width + this.sprite_width/2)
			this.destroy();
    }
}

TestScene({
	onStart: () => {
		Timer.every(1000,()=>{
			new Ball();
		});
	}
})

File: projects/dodgem/scripts/ball.lua
local SOCCER,BEACH,SPIKE,BOWLING,SMILE = 1,2,3,4,5
local VIEW_BALL_RANGE = false

Ball = Entity("Ball",{
	hitbox=true,
	reaction="cross",
	animations={ "ball" },
	align="center",
	spawn=function(self)
		self.anim_frame = Math.random(1,5)
			
		--				soccer, 	beach, 		spike, 		bowling, 	smile
		local scaling = {0.75,		1,			0.5,		1.5,		0.3 }
		local radius = 	{35,		50,			25,			75,			18 }
		local sound =	{'switch',	'car_door',	'crunch',	'cannon',	'drip'}
		
		self.scale = scaling[self.anim_frame]
		self.radius = radius[self.anim_frame]
		
		-- smile: bounce on floor and ceiling
		if self.anim_frame == SMILE then
			self.x = Game.width + self.radius
			self.y = Game.height/2
			self.hspeed = -100
			self.gravity = 20
			self.gravity_direction = table.random{90,270}
		end
		
		-- spike: move slowly vertically
		if self.anim_frame == SPIKE then 
			self.x = Math.random(self.width, Game.width - self.width)
			self.y = table.random{-self.radius, Game.height + self.radius}
			self.vspeed = 200 * -Math.sign(self.y)
		end
		
		-- soccer/beach/bowling: move left to right
		if self.anim_frame == SOCCER or self.anim_frame == BEACH or self.anim_frame == BOWLING then 
			self.x = -self.radius
		end
		
		-- soccer: move left to right at slight angle
		if self.anim_frame == SOCCER then 
			self.vspeed = 20 * table.random{-1,1}
			self.hspeed = 100
			self.y = Math.random(self.height, Game.height - self.height)
		end
		
		-- bowling: move SLOWLY left to right at slight angle
		if self.anim_frame == BOWLING then 
			self.vspeed = 20 * table.random{-1,1}
			self.hspeed = 40
			self.y = Math.random(self.height, Game.height - self.height)
		end
		
		-- beach
		if self.anim_frame == BEACH then 
			self.hspeed = 80
			self.percent = Math.random(0,100)
		end
		
		-- bowling
	end,
	is_colliding = function(self, other)
		return Math.distance(other.x, other.y, self.x, self.y) <= self.radius
	end,
	update = function(self, dt)
		-- smile/soccer/bowling: bounce off bottom of screen
		if (self.anim_frame == SMILE or self.anim_frame == SOCCER or self.anim_frame == BOWLING) 
		and (self.y + self.radius > Game.height or self.y - self.radius < 0) then 
			self.vspeed = -self.vspeed
		end
		-- ball: move in sine wave
		if self.anim_frame == BEACH then 
			self.y = Math.sinusoidal(self.height, Game.height - self.height, nil, self.percent/100)
		end
		-- NOT spike: die if off the screen
		if self.anim_frame ~= SPIKE and (self.x - self.width > Game.width or self.x + self.width < 0 or self.y - self.height > Game.height or self.y + self.height < 0) then
			self:destroy()
		end
		-- spike: die if off top/bottom of the screen
		if self.anim_frame == SPIKE and(self.y - self.height > Game.height or self.y + self.height < 0) then 
			self:destroy()
		end
	end,
	draw = function(self, d) 
		d()
		
		if VIEW_BALL_RANGE and player then
			Draw{
				{ 'color', 'green'},
				{ 'scale', 1/self.scale },
				{ 'print', Math.floor(Math.distance(player.x, player.y, self.x, self.y)) },
				{ 'color', 'green', 0.3 },
				{ 'circle', 'fill', 0, 0, self.radius }
			}
		end
	end
})